### PR TITLE
crypto: Expose a method to pin a user's identity

### PR DESCRIPTION
**Pull request by andybalaam**
Part of https://github.com/element-hq/element-meta/issues/2526

### CODE REVIEW

File: crates/matrix-sdk/src/room/mod.rs [hunk 1]
```diff
@@ -403,6 +403,18 @@ impl Room {
...
+    /// identity before.
+    #[cfg(feature = "e2e-encryption")]
+    pub async fn pin_user_identity(&self, user_id: &UserId) -> Result<()> {
+        let x = self.client.encryption().get_user_identity(user_id).await?;
+
+        match x {
+            Some(x) => Ok(x.pin().await?),
+            None => Err(Error::UserNotFound),
+        }
+    }
```
Comment (richvdh) [inline crates/matrix-sdk/src/room/mod.rs:416]: this doesn't seem specific to rooms... it seems a bit odd to expose it as part of `Room`?

File: crates/matrix-sdk/src/room/mod.rs [hunk 2]
```diff
@@ -403,6 +403,18 @@ impl Room {
         IdentityStatusChanges::create_stream(self.clone()).await
     }
 
+    /// Mark this user as "pinned" meaning that we have seen their cryptographic
+    /// identity before.
+    #[cfg(feature = "e2e-encryption")]
+    pub async fn pin_user_identity(&self, user_id: &UserId) -> Result<()> {
+        let x = self.client.encryption().get_user_identity(user_id).await?;
+
+        match x {
```
Comment (richvdh) [inline crates/matrix-sdk/src/room/mod.rs:412]: nit: either inline `x`, or give it a better name?

File: bindings/matrix-sdk-ffi/src/room.rs [hunk 3]
```diff
@@ -609,6 +609,10 @@ impl Room {
         })))
     }
 
+    pub fn pin_user_identity(&self, user_id: &str) -> Result<(), ClientError> {
```
Comment (richvdh) [inline bindings/matrix-sdk-ffi/src/room.rs:612]: could do with some documentation?

File: crates/matrix-sdk/src/room/mod.rs (hunk 2 above)
Comment (andybalaam) [reply crates/matrix-sdk/src/room/mod.rs]: Thanks, sorry I realised this the same time as you did and force-pushed an update.

File: bindings/matrix-sdk-ffi/src/room.rs (hunk 3 above)
Comment (andybalaam) [reply bindings/matrix-sdk-ffi/src/room.rs:612]: We don't document in the FFI layer normally - is there some reason why this one requires it?

File: crates/matrix-sdk/src/room/mod.rs (hunk 1 above)
Comment (andybalaam) [reply crates/matrix-sdk/src/room/mod.rs]: This was particularly requested by @stefanceriu - I guess it's for convenience because they have a room already available. I guess we could put it on `Client`. Any comment @stefanceriu ?

Comment (dkasak) [reply crates/matrix-sdk/src/room/mod.rs]: Agreed this doesn't make sense to be on `Room`. Please let's not make the API mess solely for the purpose of convenience. `Client` does make sense.

Comment (stefanceriu) [reply crates/matrix-sdk/src/room/mod.rs]: Yes, that was the reason. Also it makes it consistent with the other verification withdrawal methods. Otherwise you could even put it in the Client or even Encryption. Up to you really ¯\_(ツ)_/¯

File: crates/matrix-sdk/src/room/mod.rs [hunk 4]
```diff
@@ -403,6 +403,20 @@ impl Room {
         IdentityStatusChanges::create_stream(self.clone()).await
     }
 
+    /// Mark this user as "pinned" meaning that we have seen their cryptographic
+    /// identity before.
+    #[cfg(feature = "e2e-encryption")]
+    pub async fn pin_user_identity(&self, user_id: &UserId) -> Result<()> {
```
Comment (dkasak) [inline crates/matrix-sdk/src/room/mod.rs:409]: We're not saying *which* identity to pin, only which user's identity should be pinned, so this feels a bit smelly. I'm not sure how we will be using this API, but is there potential for a switcheroo/race here?

We observe one identity for some user U, which we want to pin. But then it switches under us, which we don't notice because the API is simply "pin *some* identity for user U, whichever one you find".

File: bindings/matrix-sdk-ffi/src/room.rs (hunk 3 above)
Comment (stefanceriu) [reply bindings/matrix-sdk-ffi/src/room.rs]: Actually we should (and have) because they get properly forwarded to the bindings.

File: crates/matrix-sdk-crypto/src/identities/user.rs [hunk 5]
```diff
@@ -125,6 +125,22 @@ impl UserIdentity {
...
+    ///
+    /// Resets this identity to the same "pinned" state assigned the first time
+    /// a user is encountered, where we follow "trust on first use" and
+    /// consider it trusted.
+    pub async fn pin(&self) -> Result<(), CryptoStoreError> {
+        match self {
+            UserIdentity::Own(_) => {
+                // Nothing to be done for our own identity: we already
+                // consider it trusted in this sense.
+                Ok(())
```
Comment (BillCarsonFr) [inline crates/matrix-sdk-crypto/src/identities/user.rs:138]: Right, it's more that we consider that you should always verifiy your own identity. We need to clarify this issue with own identity as it is surfacing in several places of the API
We could also error? or maybe one should do `identity.other().map(.. pin..)`?

**Review by BillCarsonFr — APPROVED** [review]
A bit original to have the API on room, but looks like it's convenient for apps

File: bindings/matrix-sdk-ffi/CHANGELOG.md [hunk 6]
```diff
@@ -31,4 +31,5 @@ Breaking changes:
 
 Additions:
 
+- Add `Room::pin_user_identity`
```
Comment (poljar) [inline bindings/matrix-sdk-ffi/CHANGELOG.md:34]: I'm sorry, but this can't be exposed in this manner. Please expose the `UserIdentity` and the `pin()` method the same way we have it in the main crate.

File: crates/matrix-sdk/src/room/mod.rs (hunk 4 above)
Comment (poljar) [reply crates/matrix-sdk/src/room/mod.rs:409]: Even if we are sure that a switcheroo can't happen in a dangerous manner at this point in time, and in the way EX wants to use it right now, things might change in the future and we have other users of the bindings. We can't guarantee that you can't shoot yourself in the foot this way.

There is a reason why this isn't exposed on the room in the main crate and the API for this should be kept 1:1.

**Review by poljar — CHANGES_REQUESTED** [review]

File: crates/matrix-sdk-crypto/src/identities/user.rs [hunk 7]
```diff
@@ -125,6 +125,22 @@ impl UserIdentity {
         }
     }
 
+    /// Consider this identity to be trusted.
+    ///
+    /// Resets this identity to the same "pinned" state assigned the first time
+    /// a user is encountered, where we follow "trust on first use" and
+    /// consider it trusted.
```
Comment (dkasak) [inline crates/matrix-sdk-crypto/src/identities/user.rs:132]: I must admit I don't understand this explanation at all.

The first thing that throws me is the word "trusted". Do we still retain the ability to differentiate between identities which are simply pinned versus properly verified?

Furthermore, the word "resets" implies to me that we will force the "pinned" state, replacing any previous state, even if it was "verified". Is this the case?

File: crates/matrix-sdk/src/room/mod.rs (hunk 4 above)
Comment (dkasak) [reply crates/matrix-sdk/src/room/mod.rs:409]: Sorry, unresolving, since I didn't see poljar's comment before hitting resolve (the UI didn't update).

File: bindings/matrix-sdk-ffi/src/room.rs (hunk 3 above)
Comment (andybalaam) [reply bindings/matrix-sdk-ffi/src/room.rs]: OK, update force-pushed with docs included

File: crates/matrix-sdk/src/room/mod.rs (hunk 4 above)
Comment (andybalaam) [reply crates/matrix-sdk/src/room/mod.rs]: Hmm, good point. I've re-structured the interface and force-pushed. Now we can get the identity first, and pin it later.

We don't currently have a good way of actually identifying which identity we have, but I think that is outside the scope of this change. (I suspect we should expose the master key?)

File: bindings/matrix-sdk-ffi/CHANGELOG.md (hunk 6 above)
Comment (andybalaam) [reply bindings/matrix-sdk-ffi/CHANGELOG.md]: Done in the force-pushed version I just did

File: crates/matrix-sdk/src/room/mod.rs (hunk 1 above)
Comment (andybalaam) [reply crates/matrix-sdk/src/room/mod.rs]: Changed to `Encryption::get_user_identity` then `UserIdentity::pin` in the force-pushed update I just did.

File: crates/matrix-sdk-crypto/src/identities/user.rs (hunk 5 above)
Comment (andybalaam) [reply crates/matrix-sdk-crypto/src/identities/user.rs:138]: Do you think we should error? I think probably not but am persuadable.

File: crates/matrix-sdk-crypto/src/identities/user.rs (hunk 7 above)
Comment (andybalaam) [reply crates/matrix-sdk-crypto/src/identities/user.rs:132]: Yes, this is very confusing at the moment! How about this wording:

```
/// "Pin" this identity, meaning it is not in "pin violation".
///
/// When we first see a user, we assume their cryptographic identity is correct.
/// We call this "pinning" the identity. If they later change their identity, then they are
/// in "pin violation": we require user interaction before they are pinned again.
///
/// Note: this is not related to verification of users, which is a more secure way of
/// confirming that the identity you are communicating with is really the person you
/// think it is.
```

Comment (andybalaam) [reply crates/matrix-sdk-crypto/src/identities/user.rs]: (Applied in 2d89f6cab4 but definitely up for further discussion)

File: crates/matrix-sdk-crypto/src/identities/user.rs (hunk 5 above)
Comment (BillCarsonFr) [reply crates/matrix-sdk-crypto/src/identities/user.rs:143]: I don't know it makes me wonder if we should pin our own identity also :/ for the transition mode. Anyhow not blocking.

File: crates/matrix-sdk/src/room/mod.rs (hunk 4 above)
Comment (BillCarsonFr) [reply crates/matrix-sdk/src/room/mod.rs]: Sry @poljar I am a bit lost. Why would keeping the API 1:1 will help for the "switcheroo" thing?
If I understand well, what would fix it is that the `pin` method takes a "master_key" as parameter and that the room identity state changes stream gives you the `master_key` to confirm that you can then later pass back as a parameter.

But even that doesn't really help the race scenario in case of 2 consecutive changes? Not sure the current pinning mecanism works like that.

Comment (poljar) [reply crates/matrix-sdk/src/room/mod.rs]: > If I understand well, what would fix it is that the pin method takes a "master_key" as parameter 

The `pin()` method takes the master key as an argument, it's just that the master key is inside of `self.inner.master_key`.

So yes, you are correct, the master key needs to be part of the function call.

> But even that doesn't really help the race scenario in case of 2 consecutive changes? Not sure the current pinning mecanism works like that.

The changes can happen at any point in time, but you have a copy of the identity which will not change under your feet while you decide if you want to pin it or not.

File: crates/matrix-sdk-crypto/src/identities/user.rs (hunk 7 above)
Comment (dkasak) [reply crates/matrix-sdk-crypto/src/identities/user.rs]:
```rust
/// Remember this identity, ensuring it does not result in a pin violation.
///
/// When we first see a user, we assume their cryptographic identity has not
/// been tampered with by the homeserver or another entity with
/// man-in-the-middle capabilities. We remember this identity and call this
/// action "pinning".
///
/// If the identity presented for the user changes later on, the newly presented
/// identity is considered to be in "pin violation". This method explicitly
/// accepts the new identity, allowing it to replace the previously pinned one
/// and bringing it out of pin violation.
///
/// An identity which is not verified and is in pin violation will require user
/// approval before it can be used. See
/// [`OtherUserIdentity::identity_needs_user_approval`].
```

How about this instead?

I aimed to avoid glossing over verification as an entirely separate matter, instead linking pinning and verification to show how they jointly impact the system. I also spell out what "correct" means when talking about an identity.

File: crates/matrix-sdk/src/room/mod.rs (hunk 4 above)
Comment (andybalaam) [reply crates/matrix-sdk/src/room/mod.rs]: @poljar does the new version match what you expected?

File: bindings/matrix-sdk-ffi/src/encryption.rs [hunk 8]
```diff
@@ -410,6 +410,50 @@ impl Encryption {
...
+    /// If the identity presented for the user changes later on, the newly
+    /// presented identity is considered to be in "pin violation". This
+    /// method explicitly accepts the new identity, allowing it to replace
+    /// the previously pinned one and bringing it out of pin violation.
+    ///
+    /// An identity which is not verified and is in pin violation will require
+    /// user approval before it can be used.
+    pub(crate) async fn pin(&self) -> Result<(), ClientError> {
+        Ok(self.inner.pin().await?)
+    }
```
[the remark this answers is not in the document]
Comment (andybalaam) [reply bindings/matrix-sdk-ffi/src/encryption.rs:456]: Done in 7355cb46da

[the remark this answers is not in the document]
Comment (andybalaam) [reply bindings/matrix-sdk-ffi/src/encryption.rs:456]: @poljar sorry, but I'd like you to take one last look at that please -  I called `get_first_key` and then `to_base64` and am not sure either of those is correct.

File: crates/matrix-sdk-crypto/src/identities/user.rs [hunk 9]
```diff
@@ -737,7 +763,21 @@ impl OtherUserIdentityData {
...
+    /// man-in-the-middle capabilities. We remember this identity and call this
+    /// action "pinning".
+    ///
+    /// If the identity presented for the user changes later on, the newly
+    /// presented identity is considered to be in "pin violation". This
+    /// method explicitly accepts the new identity, allowing it to replace
+    /// the previously pinned one and bringing it out of pin violation.
+    ///
+    /// An identity which is not verified and is in pin violation will require
+    /// user approval before it can be used. See
```
[the remark this answers is not in the document]
Comment (dkasak) [reply crates/matrix-sdk-crypto/src/identities/user.rs:779]: Yep, we should, though I'm out of my depth here since I've not been keeping this close of an eye on the new trust requirement code.

[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply crates/matrix-sdk-crypto/src/identities/user.rs:779]: We don't require the violation to be resolved before using the identity anymore (using here is for sharing room key or trust requirement for decryption). It was considered but it is per design a non blocking warning, so the sdk remembers if the warning was displayed to the user, until resolved the warning should be shown by UI.
So just outdated doc?

[the remark this answers is not in the document]
Comment (dkasak) [reply crates/matrix-sdk-crypto/src/identities/user.rs:779]: Hmm, I have a number of questions now.

- So what exactly is the effect of `identity_needs_user_approval() == true`? It simply displays a warning with no other effect?
- Do we differentiate between a `identity_needs_user_approval() == true` state created due to a verified identity mismatch vs a pin violation? Or is the effect exactly the same in both cases?
- Is any of this going to change after the "transition period"?

Also, can you suggest wording that would make it accurate?

File: bindings/matrix-sdk-ffi/src/encryption.rs [hunk 10]
```diff
@@ -410,6 +410,63 @@ impl Encryption {
...
+
+    /// Get the public part of the Master key of this user identity.
+    ///
+    /// The public part of the Master key is usually used to uniquely identify
+    /// the identity.
+    pub(crate) fn master_key(&self) -> Result<String, ClientError> {
```
Comment (poljar) [inline bindings/matrix-sdk-ffi/src/encryption.rs:462]: This is fine from my point of view, though usually we return an `Result<Option<T>>` for these cases since it's easier to distinguish the "Missing master key" case from other error types. This is especially true for `ClientError` which has only a single variant and people would need to match on the string.

The same applies to the `get_user_identity()` method.

File: crates/matrix-sdk-crypto/src/identities/user.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (andybalaam) [reply crates/matrix-sdk-crypto/src/identities/user.rs]: > What exactly is the effect of identity_needs_user_approval() == true? It simply displays a warning with no other effect?

Yes. (This is for _unverified_ identities.)

> So what is the behaviour when a verified identity is mismatched?

In legacy/transition mode (they are the same) we will block sending and decrypting messages to/from an identity with a verification violation. I think it will be the same in future too.

> what is the name of the method which enforces this case?

`withdraw_verification` changes a (potentially-violating) identity back into a pinned identity. `has_verification_violation` asks whether this identity has a verification violation.

I find these methods very confusing and am trying to think about how to structure them better.

> Is any of this going to change after the "transition period"?

I don't think so.

> Also, can you suggest wording that would make it accurate?

I went for "UIs should display a warning to the user when encountering an identity which is not verified and is in pin violation."

File: bindings/matrix-sdk-ffi/src/encryption.rs (hunk 10 above)
Comment (andybalaam) [reply bindings/matrix-sdk-ffi/src/encryption.rs]: Fixed in 562316c4e5
